enterprise/server/backends/pebble_cache/pebble_cache.go

```diff
@@ -1631,9 +1631,19 @@ func (p *PebbleCache) Get(ctx context.Context, r *rspb.ResourceName) ([]byte, er
 		return nil, err
 	}
 	defer rc.Close()
-	var buffer bytes.Buffer
-	_, err = io.Copy(&buffer, rc)
-	return buffer.Bytes(), err
+	buf := bytes.NewBuffer(make([]byte, 0, bufferSize(r)))
+	_, err = io.Copy(buf, rc)
+	return buf.Bytes(), err
+}
+
+func bufferSize(r *rspb.ResourceName) int {
+	if r.GetCacheType() != rspb.CacheType_CAS {
+		// The median and average AC results are less than 4KiB: go/action-result-size
+		return 4 * 1024
+	}
+	// Clamp the size between 0 and 100MB, to protect from invalid and
+	// malicious requests.
+	return min(100_000_000, max(0, int(r.GetDigest().GetSizeBytes())))
 }
```

Comment:
this is a little scary -- we'll need to test but it feels like it could easily over allocate a lot of memory? Could we use bytebufferpool somehow here?